crates/driver/src/domain/competition/auction.rs

```diff
@@ -23,26 +34,52 @@ pub struct Auction {
 }
 
 impl Auction {
-    pub fn new(
+    pub async fn new(
         id: Option<Id>,
-        mut orders: Vec<competition::Order>,
+        orders: Vec<competition::Order>,
         tokens: impl Iterator<Item = Token>,
-        gas_price: eth::GasPrice,
         deadline: Deadline,
-        weth: eth::WethAddress,
-    ) -> Result<Self, InvalidTokens> {
+        eth: &Ethereum,
+    ) -> Result<Self, Error> {
         let tokens = Tokens(tokens.map(|token| (token.address, token)).collect());
 
         // Ensure that tokens are included for each order.
+        let weth = eth.contracts().weth_address();
         if !orders.iter().all(|order| {
             tokens.0.contains_key(&order.buy.token.wrap(weth))
                 && tokens.0.contains_key(&order.sell.token.wrap(weth))
         }) {
-            return Err(InvalidTokens);
+            return Err(Error::InvalidTokens);
```

Comment:
nit: may be good to at least log which tokens are missing (I imagine this being otherwise very complex to debug)

Reply:
Personally, I'm OK with no log in the long run:
- We get a `InvalidTokens` error
- We look at the stored auction JSON from the autopilot.

Now, that being said - I don't think we are saving the auction JSON that is being sent to the drivers for every run. I really think we should in the context of co-location (as it would want to store everything we send to externally operated `driver`s). Once we do store these auction JSON documents, however, I think the extra log here isn't super useful (you will always check the auction JSON anyway to check if the missing token is indeed missing, so I don't think it saves you much time).